README.md
# Stock Gameserver

Dividends are ignored.

Data provided for free by [IEX](https://iextrading.com/developer). View [IEX’s Terms of Use](https://iextrading.com/api-exhibit-a/).

## Authentication

To authenticate requests, you must first create an account at the `/api/v1/account` endpoint. You can create as many accounts as you want to.

Authorization HTTP header example:
* `Authorization: Basic M3JkcGFydHlfY2xpZW50aWQ6amtmb3B3a21pZjkwZTB3b21rZXBvd2U5aXJram8zcDlta2Z3ZQ==`

Basic access authentication is used, with the user ID as the username and secret as the password. The header should be the string `Basic ` plus the Base64 encoded string {user_id}:{secret}.
It is very important to make sure there is no whitespace around or in the string that is being Base64 encoded. If in doubt, try to Base64 encode the sample values provided below and compare them to the results:
* Example user ID: `3rdparty_clientid`
* Example secret: `jkfopwkmif90e0womkepowe9irkjo3p9mkfwe`
* Concatenated to become: `3rdparty_clientid:jkfopwkmif90e0womkepowe9irkjo3p9mkfwe`
* Base64 encoded to: `M3JkcGFydHlfY2xpZW50aWQ6amtmb3B3a21pZjkwZTB3b21rZXBvd2U5aXJram8zcDlta2Z3ZQ==`
* Resulting in the Authorization header: `Basic M3JkcGFydHlfY2xpZW50aWQ6amtmb3B3a21pZjkwZTB3b21rZXBvd2U5aXJram8zcDlta2Z3ZQ==`

Endpoints which require authentication are marked as such in the description below.

If the authentication failed, the status code `401` will be returned.

## API

The api uses the `/api/vX` versioning schema where `X` is a number. New versions will be created for breaking changes. New endpoints, deprecations and removals will be announced/scheduled through the CHANGELOG.md file.
There is also a `X-STOCKS-DEPRECATION: Date` header which contains the date on which the endpoint will be removed.

All endpoints return json.

### GET /isalive/

Returns a 200 if alive.

### GET /api/v1/quote/{symbol}/

Returns a quote for the given symbol if the according information could be retrieved. All stocks from https://iextrading.com/apps/stocks are supported, feel free to use their API for your purposes.

Example:
```json
{
    "symbol": "AAPL",
    "latestPrice": 187.59
}
```

### POST /api/v1/account/{name}/

Create an account with the given name and an initial balance of 10000. Please use your real name. Names are not unique, so you can create as many accounts as you want.

No request body.

Returns the userId and a secret that you must remember for further calls. You have to start over if you lose them.

Example:
```json
{
    "userId": "97asndyf9-sa90d7fnsad",
    "secret": "as9-8dn8g08basnfd-asgdfsdfihsdf",
    "name": "Michael Bahr"
}
```

_If you lose those credentials, then there is no way to retrieve them._

### GET /api/v1/balance/

**Requires authentication**, see chapter on Authentication.

Returns your balance.

Example:
```json
{
    "userId": "97asndyf9-sa90d7fnsad",
    "amount": 10000.0
}
```

### GET /api/v1/transactions/

**Requires authentication**, see chapter on Authentication.

Returns a list of all your transactions. Transactions are comleted buy or sell orders.

Example:
```json
[
    {
        "id": "0asb7dfy087asbdf808sadfbs",
        "userId": "97asndyf9-sa90d7fnsad",
        "pricePerUnit": 175.87,
        "amount": 10,
        "symbol": "AAPL",
        "isBuy": true,
        "instant": "2018-09-28T19:16:07.496Z"
    }, ...
]
```

### GET /api/v1/transactions/{symbol}/

**Requires authentication**, see chapter on Authentication.

Same as above, but filtered by symbol.

Example:
```json
[
    {
        "id": "0asb7dfy087asbdf808sadfbs",
        "userId": "97asndyf9-sa90d7fnsad",
        "pricePerUnit": 175.87,
        "amount": 10,
        "symbol": "AAPL",
        "isBuy": true,
        "instant": "2018-09-28T19:16:07.496Z"
    }, ...
]
```

### POST /api/v1/order/

**Requires authentication**, see chapter on Authentication.

An order may either succeed and result in a transaction or fail with an error code and an error message. Contrary to stock exchanges, orders are processed immediately and have no delayable aspect.

Example Request:
```json
{
    "symbol": "AAPL",
    "isBuy": true,
    "amount": 10
}
```

Example Response:
```json
{
    "id": "0asb7dfy087asbdf808sadfbs",
    "userId": "97asndyf9-sa90d7fnsad",
    "pricePerUnit": 175.87,
    "amount": 10,
    "symbol": "AAPL",
    "isBuy": true,
    "instant": "2018-09-28T19:16:07.496Z"
}
```

Repeated calls will try to place multiple orders.

#### Possible Error Codes

Error messages are provided in the body.

`400`: Insufficient Funds. You don't have the required <amount>

`400`: Insufficient Shares. For the given symbol you only have <amount> shares.

`404`: The symbol <symbol> could not be found. If it's not a typo, then please report this to the admin.

### GET /api/v1/analyses/

Returns a list of analyses as seen on [finanzen.net](https://www.finanzen.net/analysen/) if that stock has a symbol (open a report and look under the headline).

`v1` returns all analyses that were tracked since the app's go-live. Futures versions may have date filters.

Example Response:
```json
[
    {
        "symbol": "NKE",
        "buy": 14,
        "hold": 7,
        "sell": 1,
        "source": "https://www.finanzen.net/analyse/Nike_buy-Stifel_Nicolaus__Co__Inc__647337",
        "instant": "2018-09-28T19:16:07.496Z"
    }, ...
]
```


CHANGELOG.md
# Changelog

## 2018-09-29

**Breaking change**

The following routes' paths were updated.

* `/api/v1/isalive` --> `/isalive/` 
* `/api/v1/quote/{symbol}` --> `/api/v1/quote/{symbol}/`
* `/api/v1/account/{name}` --> `/api/v1/account/{name}/`
* `/api/v1/balance/{userId}` --> `/api/v1/balance/`
* `/api/v1/transactions/{userId` --> `/api/v1/transactions/`
* `/api/v1/transactions/{userId}/{symbol}` --> `/api/v1/transactions/{symbol}/`
* `/api/v1/order/{userId}` --> `/api/v1/order/`
* `/api/v1/analyses` --> `/api/v1/analyses/`

## 2018-06-29

All routes were extended with the `v1` version to enable later versioning of routes.

**NEW**

* `/api/v1/isalive`
* `/api/v1/quote/{symbol}`
* `/api/v1/account/{name}`
* `/api/v1/balance/{userId}`
* `/api/v1/transactions/{userId`
* `/api/v1/transactions/{userId}/{symbol}`
* `/api/v1/order/{userId}`
* `/api/v1/analyses`


src/main/java/de/hackerstolz/stockgameserver/web/OrderResource.java
package de.hackerstolz.stockgameserver.web;

import de.hackerstolz.stockgameserver.config.SecurityUtils;
import de.hackerstolz.stockgameserver.exception.InsufficientFundsException;
import de.hackerstolz.stockgameserver.exception.InsufficientSharesException;
import de.hackerstolz.stockgameserver.exception.StockNotFoundException;
import de.hackerstolz.stockgameserver.model.Order;
import de.hackerstolz.stockgameserver.service.OrderService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/order/")
public class OrderResource {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private final OrderService orderService;

    public OrderResource(OrderService orderService) {
        this.orderService = orderService;
    }

    @PostMapping
    public ResponseEntity<?> placeOrder(@RequestBody final Order order) {
        if (order.getIsBuy() == null || order.getAmount() == null || order.getSymbol() == null) {
            log.info("Invalid order: {}", order);
            return ResponseEntity.badRequest().build();
        }

        try {
            final String userId = SecurityUtils.getCurrentUserLoginAsString();
            return ResponseEntity.ok(orderService.placeOrder(order, userId));
        } catch (final StockNotFoundException e) {
            return ResponseEntity.status(404).body("The symbol " + e.getSymbol() + " could not be found. If it's not a "
                    + "typo, then please report this to the admin.");
        } catch (final InsufficientFundsException e) {
            return ResponseEntity.status(400).body(
                    "Insufficient Funds. You don't have the required " + e.getRequiredAmount().intValue());
        } catch (final InsufficientSharesException e) {
            return ResponseEntity.status(400).body(
                    "Insufficient Shares. For the given symbol you only have " + e.getAvailableShares() + " shares");
        }
    }
}
